### ISSUE TITLE
Expose stats store singleton via Api::Api

### ISSUE THREAD

**Issue by ahedberg** [envoyproxy/envoy#6706]
*Title*: Expose stats store singleton via Api::Api

*Description*: We have a use case where we want to read stats exported by the xDS APIs from a filter. The stats store isn't globally accessible right now, and we'd rather not make admin requests to /stats from the filter. Could we expose the stats store via Api::Api, thus making it accessible to the filter via FactoryContext?

**Comment by junr03** [comment]
@jmarantz I think you are the best person to comment on this probably.

**Comment by mattklein123** [comment]
I would be OK exposing the root scope (not store) as a const object only. @jmarantz WDYT?

**Comment by jmarantz** [comment]
Root Scope SGTM. Store inherits from Scope and adds the counters(), gauges(), and histograms() APIs . Ashley -- is that sufficient? You would just look up stats you know the name of?

**Comment by mattklein123** [comment]
Oh yeah I guess if it needs to list out all stats, doing the store might be needed. As long as it's const sounds OK.

**Comment by ahedberg** [comment]
We know the names of the stats we care about, so root scope should be sufficient.

**Comment by ahedberg** [comment]
While working on a test in #6712, I realized that the counter/gauge/histogram APIs in Stats::Scope are not const. I'm not sure it's reasonable to affix the const qualifier to those member functions--Envoy::Stats::IsolatedStoreImpl is an eventual subclass, and does decidedly non-const things in its implementations (like adding cached stats to a hash map). This unfortunately makes a const Stats::Scope& not particularly useful.

Thoughts?

**Comment by mattklein123** [comment]
@ahedberg it would be more work, but one option would be to add a const get method to the scope which will only get a stat and not create it. This would probably be generally useful. I don't think we should provide a non-const root scope to all extensions. The other option would be to just return the const store (not scope) which allows you to iterate over all stats. Maybe this would be enough? cc @jmarantz for thoughts.

**Comment by jmarantz** [comment]
Having const getters sgtm. Now; should the returned Counter& also be const, so that you can observe the value but not increment?

What's the usage model for const stats?

**Comment by jmarantz** [comment]
Sorry I didn't read Matt's and Ashley's comments closely enough and missed a key point -- I think if we want to provide a conceptually const stat store or scope, that provides read-only name lookups, it should still have the same caching behavior and we should annotate the hash tables as mutable, adding new behavior that will avoid creating new stats for a new name, but will populate the caches as needed.  

The new method would probably return pointers rather than references, as a non-existing name has to have some behavior. My personal bias is to prefer that to raising exceptions, but if we decide to make the getCounter() method return a ref or throw, we could always wrap it at Google.

I don't have a strong feeling about whether we need to provide the versions that return a vector of counters, in the API.

**Comment by ahedberg** [comment]
I think adding `Counter* const getCounter(const std::string& name) const;` and friends for gauges and histograms to the scope, which will populate the underlying cache but not create new stats, makes sense and is worth investing time in. Otherwise I suspect we'll see folks reinventing the "iterate over all counters to find the one(s) we care about" loop.

I'll move forward with that plan if that sounds good to you.

**Comment by jmarantz** [comment]
SGTM overall, modulo 2 things:
 * can you make the implementation take a Stats::StatName rather than a const std::string& name? All new calls should use StatName.
 * Note that the implementation of this in thread_local_store.cc will likely merge-conflict with @fredlas #5910 . That's not a reason not to do it, we should just be aware of it. Hopefully that will merge before too long.

**Comment by ahedberg** [comment]
So, fun discovery after a few days of debugging: `BaseIntegrationTest` has an `Api` object that it uses for various things, including constructing an `IntegrationTestServerImpl`. However, the `IntegrationTestServerImpl` creates its own stats store, symbol table, etc., completely ignoring whatever its `Api` was constructed with. This is fine...until you're writing an integration test and are trying to read stats out of the `Api`'s scope, only to find that they're stored somewhere else.

tl;dr: Yak-shaving continues, but I haven't dropped this.
